### DISCUSSION TITLE
[Feature]: Support creating containers on-the-fly using CloudNative buildpacks

### DISCUSSION THREAD

**Discussion by hariohmprasath** [testcontainers/testcontainers-java#5864] [Ideas]
### Module

Core

### Problem

Currently, as a developer I can create temporary containers on the fly from a docker file using `ImageFromDockerfile` class. But what if I don’t have a `Dockerfile` and I just want to point to a codebase and generate a container image out of it on the fly and use the same container to run some integration tests? Currently, it’s not possible. Here are some scenarios when a docker file is not available during the initial development phase:

* In an enterprise setup `Dockerfile` is created and maintained by DevOps/SecOps team so as a developer I have to wait till I get a formally approved docker file before I run any tests using testcontainers

* If I have a legacy application, then creating a docker file will require much more efforts from the team as they have to research on the supported runtime, dependencies, etc before building one

### Solution

Add support CloudNative [buildpacks](https://buildpacks.io/)

### Benefit

By supporting buildpacks users don’t need to create or manage docker files. After buildpack integration with `testcontainers` users can point the source code and generate containers on-the-fly without a docker file. 

Example:

```
ImageFromDockerfile.withBuildPack(‘<folder_path>‘,’<builder>‘)
```

This is one of the most popular use case that most enterprise uses to quickly modernize their applications. @eddumelendez let me know teams thoughts on this as always I would love to contribute if we approve the proposal.

### Alternatives

The user have to create a `Dockerfile` and then use it to create an image with `testcontainers`.

### Would you like to help contributing this feature?

Yes

**Comment by eddumelendez** [comment]
@hariohmprasath thanks for the idea! I think you may be interested in #5623. Although it is not related to buildpacks it covers the same need. I think it would be great to have something similar to what the spring boot team which AFAIK doesn't use pack cli but has its own implementation. Unfortunately, it's not something we can reuse 😢 

would like to hear from @kiview and @bsideup on this one

**Comment by hariohmprasath** [comment]
Hi @eddumelendez, Thanks for your response and pointing me to the PR. 

Here are my two cents: 
Based on my experience `jLib` only supports building java docker images, but users working with `testcontainers` would like to launch on-the-fly containers based on any language runtime for a variety of usecases. So falling back on existing buildpack/builder implementations like `paketo` would help us provide support on a variety of runtimes rather than us building something of our own. 

But I would love to hear feedback from maintainers and community members. Looking forward to the discussion.

**Comment by eddumelendez** [comment]
@hariohmprasath you may interested on [this](https://github.com/eddumelendez/testcontainers-buildpacks-examples/blob/main/src/test/java/com/example/testcontainersbuildpacksexamples/BuildpackTest.java) :)

**Reply by hariohmprasath** [reply]
Thanks for sharing @eddumelendez, love to see this working :)
